usrp_drivers/n200/n200_driver.cpp

```diff
@@ -24,271 +24,339 @@
+    time_diff = std::chrono::duration_cast<std::chrono::microseconds>
+                                                (ctr_freq_setup_end - ctr_freq_setup_begin).count();
+    DEBUG_MSG("TRANSMIT center frq tuning time: " << time_diff << "us");
 
 
-    std::chrono::steady_clock::time_point sample_begin = std::chrono::steady_clock::now();
+    auto sample_unpack_begin = std::chrono::steady_clock::now();
 
-    if (dp.samples_size() > 0) {  // ~700us to unpack 4x1600 samples
-      samples = make_samples(dp);
+    //Parse new samples from driver packet if they exist.
+    if (driver_packet.samples_size() > 0)
+    {  // ~700us to unpack 4x1600 samples
+      samples = make_tx_samples(driver_packet);
       samples_per_buff = samples[0].size();
       samples_set = true;
     }
 
-    std::chrono::steady_clock::time_point sample_end = std::chrono::steady_clock::now();
-    std::cout << "TRANSMIT sample unpack time: "
-      << std::chrono::duration_cast<std::chrono::microseconds>
-                                                  (sample_end - sample_begin).count()
-      << "us" << std::endl;
+    auto sample_unpack_end = std::chrono::steady_clock::now();
+
+    time_diff = std::chrono::duration_cast<std::chrono::microseconds>
+                                                  (sample_unpack_end - sample_unpack_begin).count();
+    DEBUG_MSG("TRANSMIT sample unpack time: " << time_diff << "us");
 
-    if ( (channels_set == false) ||
-      (center_freq_set == false) ||
-      (samples_set == false) ) {
+    //In order to transmit, these parameters need to be set at least once.
+    if ((usrp_channels_set == false) || (center_freq_set == false) ||(samples_set == false))
+    {
       // TODO(keith): throw error
       continue;
     }
 
 
-    std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
+    auto setup_end = std::chrono::steady_clock::now();
 
-    std::cout << "TRANSMIT Total set up time: "
-          << std::chrono::duration_cast<std::chrono::microseconds>
-                                                      (end - begin).count()
-          << "us" << std::endl;
+    time_diff = std::chrono::duration_cast<std::chrono::microseconds>
+                                                      (setup_end - setup_begin).count();
+    DEBUG_MSG("TRANSMIT Total set up time: " << time_diff << "us");
 
 
 
-    if ( (dp.sob() == true) ) {
-      //The USRP needs about a 10ms buffer into the future before time
-      //commands will correctly work
-      time_zero = usrp_d->get_usrp()->get_time_now() + uhd::time_spec_t(10e-3);
+    if (driver_packet.sob() == true)
+    {
+      //The USRP needs about a SET_TIME_COMMAND_DELAY buffer into the future before time
+      //commands will correctly work. This was found through testing and may be subject to change
+      //with USRP firmware updates.
+      time_zero = usrp_d.get_current_usrp_time() + uhd::time_spec_t(SET_TIME_COMMAND_DELAY);
```

Comment:
just an idea: have a small test program that finds the lowest value possible for SET_TIME_COMMAND_DELAY automagically (maybe connect one gpio of one usrp to another usrp's gpio?)

Reply:
We could try that. Im not sure this would work on the fly, but we could at least have a better determination of the delay.